File: ballast-core/npm/src/domain/models/board.ts
import { BoardType } from "./board-type";
import { CubicCoordinates } from "./cubic-coordinates";
import { Tile } from "./tile";
import { TileShape } from "./tile-shape";

export class Board {

    public readonly id: string;
    public readonly boardType: BoardType;
    public readonly tileShape: TileShape;
    public readonly tileMap: Map<string, Tile>;

    private cachedTiles?: Tile[];
    public get tiles(): Tile[] {
        if (!this.cachedTiles) {
            this.cachedTiles = Array.from(this.tileMap.values());
        }
        return this.cachedTiles || [];
    }

    public getTile(orderedTriple: number[]) {
        return this.tileMap.get(orderedTriple.toLocaleString());
    }

    public constructor(id: string, boardType: BoardType, tileShape: TileShape, tiles: Tile[]) {
        this.id = id;
        this.boardType = boardType;
        this.tileShape = tileShape;
        this.tileMap = new Map(this.mapTiles(tiles));
    }

    private *mapTiles(tiles: Tile[]) {
        let tileIterator = tiles;
        for (let tile of tileIterator) {
            let item: [string, Tile] = [ tile.cubicCoordinates.toOrderedTriple().toLocaleString(), tile ];
            yield item;
        }
    }

    public getRandomPassableCoordinates(): CubicCoordinates {
        let passables = this.tiles
            .filter(x => !!x.terrain.passable);
        let index = Math.floor(Math.random() * Math.floor(passables.length)) - 1;
        return passables[index].cubicCoordinates;
    }

    public getTileFromCoordinates(cubicCoordinates: CubicCoordinates) {
        return this.tiles.find(x => x.cubicCoordinates.equals(cubicCoordinates));
    }

}

File: ballast-core/dotnet/src/domain/models/Board.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ballast.Core.Domain.Models
{
    public class Board
    {

        public BoardType BoardType { get; private set; }
        public Guid Id { get; private set; }
        public IEnumerable<Tile> Tiles { get; private set; }
        public TileShape TileShape { get; private set; }

        public Board(BoardType boardType, Guid id, IEnumerable<Tile> tiles, TileShape tileShape)
        {
            BoardType = boardType;
            Id = id;
            Tiles = tiles;
            TileShape = tileShape;
        }
        
        public CubicCoordinates GetRandomPassableCoordinates()
        {
            var passables = Tiles.Where(x => x.Terrain.Passable);
            var index = new Random().Next(1, passables.Count()) - 1;
            return passables.ElementAt(index).CubicCoordinates;
        }

        public Tile GetTileFromCoordinates(CubicCoordinates cubicCoordinates) =>
            Tiles.SingleOrDefault(x => x.CubicCoordinates.Equals(cubicCoordinates));

    }
    
}